core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js

```diff
@@ -191,6 +227,100 @@ oppia.controller('LearnerDashboard', [
       return value;
     };
 
+    $scope.sortFeedbackFunction = function(entity) {
+      return entity[$scope.currentFeedbackThreadsSortType];
+    }
+
+    $scope.onClickThread = function(
+      explorationId, threadId, explorationTitle) {
+      var threadDataUrl = (
+        '/learnerdashboardthreadhandler/' + explorationId + '/' + threadId);
+      $scope.threadTitle = explorationTitle;
+      $scope.feedbackThreadActive = true;
+      $scope.explorationId = explorationId;
+      $scope.threadId = threadId;
+
+      for (var index = 0; index < $scope.threadSummaries.length; index++) {
+        if ($scope.threadSummaries[index].exploration_id === explorationId &&
+            $scope.threadSummaries[index].thread_id === threadId) {
+          $scope.threadIndex = index;
+          var threadSummary = $scope.threadSummaries[index]
+          if (typeof threadSummary.second_last_message_read !== 'undefined') {
+            $scope.threadSummaries[index].second_last_message_read = true;
+          }
+          if (threadSummary.last_message_read === false) {
+            $scope.numberOfUnreadThreads -= 1;
+          }
+          threadSummary.last_message_read = true;
+        }
+      }
+
+      $http.get(threadDataUrl).then(function(response) {
+        var data = response.data;
+        $scope.messageSummaries = data.message_summary_list;
+      });
+    };
+
+    $scope.backFromThread = function() {
+      $scope.feedbackThreadActive = false;
+    };
+
+    $scope.addNewMessage = function(explorationId, threadId, newMessage) {
+      var url = '/threadhandler/' + explorationId + '/' + threadId;
+      var payload = {
+        updated_status: null,
+        updated_subject: null,
+        text: newMessage
+      };
+      $scope.messageSendingInProgress = true;
+      $http.post(url, payload).success(function() {
+        var messageSummary = {};
```

Comment:
Sure. Take a look at e.g. [ExplorationObjectFactory](https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/domain/exploration/ExplorationObjectFactory.js), for instance, and see how it's used in the codebase. Basically, when a response comes back from the backend, that dict is converted into a domain object that then subsequently gets used in the frontend logic. It has helper methods for common operations and is basically a plain JavaScript object. In general, the frontend logic should be working with domain objects and not the raw dicts from the backend.